// nixpsit-test/general/nixpsit-test-connection.js
const http = require('http');
const api = require("./nixpsit-test-api.js");

class Connection {
	constructor() {
		this.isConnected = false;
		this.user_id;
		this.expiration_date;
		this.user_hash;
	}
	
	async connect(user, password) {
		let connectionData = await api.call("auth.login", {"user_name": user, "user_pass": password });
		if(connectionData.error === undefined) {
			this.user_id = connectionData.user_id;
			this.expiration_date = connectionData.expiration_date;
			this.user_hash = connectionData.user_hash;
			this.isConnected = true;
		} else {
			this.isConnected = false;
			throw new Error(connectionData.error);
		}
	}

	fetchJSON(pUrl) {
		if(this.isConnected === false) {
			throw new Error("Connection: please make a connection before fetching an url");
		}
		var that = this;
		return new Promise(function(resolve, reject) {
			var request = http.request(pUrl, function(pResponse) {
				let json = "";
				pResponse.setEncoding('utf8');
				pResponse.on("data", function(chunk) {
					json += chunk.toString();
				});
				pResponse.on("end", function() {
					try{
						resolve(JSON.parse(json));
					} catch(e) {
						reject(e);
					}
				})
			});
			that._setCookies(request);
			request.end();
		});
	}

	_setCookies(pRequest) {
		pRequest.setHeader("Cookie", ["user_id=" + this.user_id, "expiration_date=" + this.expiration_date, "user_hash=" + this.user_hash]);
	}
}

module.exports = new Connection();
