// nixpsit-test/general/UnitTest/types/nixpsit-unit-ObjectType.js
if(typeof require !== "undefined") {
	var TypeDefinition = require("./nixpsit-unit-TypeDefinition.js");
	var EndOfType = require("./nixpsit-unit-EndOfType.js");
}

(function() {
	ObjectType = function(pRequired, pObjectDefinition, pRandom) {
		this.typeDefinition(pRequired);
		
		this.objectDefinition = pObjectDefinition;
		this.random = pRandom;
		
		this.testObject; 
	}
	
	ObjectType.prototype = Object.assign({}, TypeDefinition.prototype, {
		constructor: ObjectType,
		typeDefinition: TypeDefinition,
		
		getValue: function(pPointer, pTestObject) {
			if(pPointer > 0) {
				this.pointer = 2;
			}
			if(JSON.stringify(this.objectDefinition) !== JSON.stringify(pTestObject.definition) || MAX_DEPTH_TEST_OBJECT > pTestObject.depth) {
				if(this.testObject === undefined) {
					var depth = pTestObject.depth;
					if(JSON.stringify(this.objectDefinition) === JSON.stringify(pTestObject.definition)) {
						depth++;
					}
					this.testObject = new TestObject(this.objectDefinition, pTestObject.allCombinations, depth);
					if(this.random === false) {
						return this.testObject.getCurrentState();
					} else {
						return this.testObject.createRandomObject();
					}
				} else {
					if(this.random === false) {
						if(this.testObject.setNextState() === true) {
							return this.testObject.getCurrentState();
						} else {
							this.testObject === undefined;
							return new EndOfType();
						}
					} else {
						return this.testObject.createRandomObject();
					}
				}
			} else {
				if(this.random === true) {
					return undefined;
				} else {
					return new EndOfType();
				}
			}
		}
	});
})()

if(typeof module !== "undefined") {
	module.exports = ObjectType;
}
