// nixpsit-test/general/UnitTest/nixpsit-unit-ObjectDefinition.js
if(typeof require !== "undefined") {
	var EndOfType = require("./types/nixpsit-unit-EndOfType.js");
}

(function() {
	ObjectDefinition = function(pPropertyDefinitions) {
		if(pPropertyDefinitions === undefined) {
			pPropertyDefinitions = [];
		}
		
		this.properties = pPropertyDefinitions;
		this.propertyPointer = 0;
	}

	ObjectDefinition.createFrom = function(pProperties) {
		var object = function() {
			this.objectDefinition(pProperties);
		}
		
		object.prototype = Object.assign({}, ObjectDefinition.prototype, {
			constructor: object,
			objectDefinition: ObjectDefinition
		});
		
		return object;
	}

	ObjectDefinition.isObjectDefinitionConstructor = function(pConstructor) {
		return !(typeof pConstructor !== "function" || typeof pConstructor.prototype.isObjectDefinition !== "function" || new pConstructor().isObjectDefinition() !== true);
	}
	
	ObjectDefinition.prototype = {
		constructor: ObjectDefinition,
		
		isObjectDefinition: function() {
			return true;
		},

		iterate: function(definition, callback) {
			for(var i = 0, len = this.properties.length; i < len; i++) {
				callback.call(null, this.properties[i]);
			}
		},
		
		getProperty: function(property) {
			for(var i = 0, len = this.properties.length; i < len; i++) {
				if(this.properties[i].name === property) {
					return this.properties[i];
				}
			}
		},
		
		next: function(pTestObject) {
			if(this.propertyPointer === this.properties.length) {
				return new EndOfType();
			}
			if(this.properties[this.propertyPointer].types.indexOf("uniqID") === -1) {
				var value = this.properties[this.propertyPointer].next(pTestObject);
				if(value instanceof EndOfType) {
					this.properties[this.propertyPointer].reset();
					this.propertyPointer++;
					return this.next(pTestObject);
				}
				return value;
			} else {
				this.propertyPointer++;
				return this.next(pTestObject);
			}
		},
		
		current: function() {
			return this.properties[this.propertyPointer];
		},
		
		reset: function() {
			this.propertyPointer = 0;
		}
	}
})()

if(typeof module !== "undefined") {
    module.exports = ObjectDefinition;
}
